internal/pkg/controller/test/tests.go

```diff
@@ -0,0 +1,296 @@
+/*
+Copyright 2021 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package test
+
+import (
+	"context"
+	"crypto/x509"
+
+	apiutil "github.com/jetstack/cert-manager/pkg/api/util"
+	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
+	cmmeta "github.com/jetstack/cert-manager/pkg/apis/meta/v1"
+	"github.com/jetstack/cert-manager/test/unit/gen"
+	. "github.com/onsi/ginkgo"
```

Comment:
question: are we using ginkgo for more than before/after functions ('suites') and `Expect` functionality?

question: do we use ginkgo in other Jetstack projects?

Reply:
I see it's in the CM e2e tests. In JSCP we use testify suite which I'd say is more idiomatic for test suite functionality. I suppose the waiting functionality is another benefit?

Not challenging it being here, but interested to know the rationale since I think something similar in testify could be easier to use for anyone coming from JSCP. ginkgo has some unconventional syntax it seems.

Reply:
@charlieegan3 

> question: are we using ginkgo for more than before/after functions ('suites') and Expect functionality?

Yeah it has some more useful things like `Eventually` etc. We are also exporting the test results to `/_artifacts` which get pushed to a google bucket in our e2e.

Optional concurrency is also nice.

> question: do we use ginkgo in other Jetstack projects?

We use it for most if not all cert-manager projects.

---

Not particularly wedded to using ginkgo, we can remove it if you feel it will become a nuisance.

Reply:
👍 thanks.